### PR TITLE
list: don't clone local repos

### PR DESCRIPTION
**Pull request by efiop**
Pre-requisite for #3590 

* [x] ❗ I have followed the [Contributing to DVC](https://dvc.org/doc/user-guide/contributing/core) checklist.

* [x] 📖 If this PR requires [documentation](https://dvc.org/doc) updates, I have created a separate PR (or issue, at least) in [dvc.org](https://github.com/iterative/dvc.org) and linked it here.

Thank you for the contribution - we'll try to review it as soon as possible. 🙏

### CODE REVIEW

File: dvc/external_repo.py [hunk 1]
```diff
@@ -59,13 +59,17 @@ def make_repo(path, **_kwargs):
...
         repo_factory=make_repo,
         **kwargs,
     )
 
+    if "subrepos" not in repo_kwargs:
+        repo_kwargs["subrepos"] = True
+
+    if "uninitialized" not in repo_kwargs:
+        repo_kwargs["uninitialized"] = True
+
```
Comment (efiop) [inline dvc/external_repo.py:72]: These should be explicit as in the ls.py below. Just keeping this as a temporary compatibility layer for other code.

File: dvc/repo/__init__.py [hunk 2]
```diff
@@ -202,6 +202,18 @@ def __init__(
...
+    @staticmethod
+    @contextmanager
+    def open(url, *args, **kwargs):
+        if os.path.isdir(url):
+            yield Repo(url, *args, **kwargs)
+            return
+
+        from dvc.external_repo import external_repo
+
+        with external_repo(url, *args, **kwargs) as repo:
```
Comment (efiop) [inline dvc/repo/__init__.py:214]: What this really is is a `Repo.clone` with a temporary dir. Having `Repo.open()` is very convenient when you just want to access the repo, no matter local or remote (in which case temp clone is totally fine).

File: dvc/repo/__init__.py [hunk 3]
```diff
@@ -202,6 +202,18 @@ def __init__(
     def __str__(self):
         return self.url or self.root_dir
 
+    @staticmethod
+    @contextmanager
+    def open(url, *args, **kwargs):
+        if os.path.isdir(url):
```
Comment (skshetry) [inline dvc/repo/__init__.py:208]: just an `exists` check would be enough here, I think.

Comment (efiop) [reply dvc/repo/__init__.py:208]: Good point, might improve the local error.

**Comment by efiop** [comment]
Ok, we'll need to revisit CLI semantics for the `list` itself (e.g. to make the docs/help make sense for `dvc list` and `dvc list path`). Merging so we could use open() in other places.

**Pull request merged**